cmd/neofs-node/main.go

```diff
@@ -152,3 +152,37 @@ func shutdown(c *cfg) {
 func (c *cfg) onShutdown(f func()) {
 	c.closers = append(c.closers, f)
 }
+
+func (c *cfg) restartMorph() error {
+	c.log.Info("restarting internal services because of RPC connection loss...")
+
+	c.shared.resetCaches()
+
+	epoch, ni, err := getNetworkState(c)
+	if err != nil {
+		return fmt.Errorf("getting network state: %w", err)
+	}
+
+	updateLocalState(c, epoch, ni)
+
+	// bootstrap node after every reconnection cause the longevity of
+	// a connection downstate is unpredictable and bootstrap TX is a
+	// way to make a heartbeat so nothing is wrong in the making sure a
+	// node is online (if it should be)
+
+	if !c.needBootstrap() || c.cfgNetmap.reBoostrapTurnedOff.Load() {
+		return nil
+	}
+
+	err = c.bootstrap()
+	if err != nil {
+		c.log.Warn("can't send re-bootstrap tx", zap.Error(err))
+	}
+
+	// drop expired sessions if any has appeared while node was sleeping
+	c.shared.privateTokenStore.RemoveOld(epoch)
```

Comment:
looks very specific, are you sure that this is and will be the only part that needs to be updated? to me `start` and `restart` aren't so different: `start` starts from "undefined" epoch while `restart` from the cached one.

Reply:
can you elaborate on your thoughts?

i look through the whole code and there is not that much code that caches smth. every other client user (in the node, i mean) just calls the methods and does everything it needs without caching

there are not that many "services" in the node in fact. storage is the main but [we cannot stop](https://github.com/nspcc-dev/neofs-node/issues/1337#issuecomment-1514791288) it, cause it would be like restarting the whole node

Reply:
i imagine any component (e.g. not yet added) which reacts the epoch change. Don't we forget to pass `epoch` to this component here? 

talking about `privateTokenStore`: we have this logical binding
https://github.com/nspcc-dev/neofs-node/blob/e5ac376c7081669fbb59c79e68cfd8eb208ccc2a/cmd/neofs-node/session.go#L49-L51

and u add new case when we handle exact epoch. Don't u think it'd be nice to have single binding of `privateTokenStore`  and new epochs?

Reply:
okay, lets leave it as it is for now

and i'd think about single epoch controller for the node as a possible enhancement (smart controller orchestrating ordinary components)